phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/tasks/CdcStreamPartitionMetadataTask.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.phoenix.schema.task.Task;
+import org.apache.phoenix.util.CDCUtil;
+import org.apache.phoenix.util.QueryUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.util.List;
+
+import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.SYSTEM_CDC_STREAM_NAME;
+import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.SYSTEM_CDC_STREAM_STATUS_NAME;
+import static org.apache.phoenix.query.QueryServices.PHOENIX_STREAMS_GET_TABLE_REGIONS_TIMEOUT;
+import static org.apache.phoenix.query.QueryServicesOptions.DEFAULT_PHOENIX_STREAMS_GET_TABLE_REGIONS_TIMEOUT;
+
+/**
+ * Task to bootstrap partition metadata when CDC is enabled on a table.
+ * Upserts one row for each region of the table into SYSTEM.CDC_STREAM and marks the status as
+ * ENABLED in SYSTEM.CDC_STREAM_STATUS.
+ */
+public class CdcStreamPartitionMetadataTask extends BaseTask  {
+
+    public static final Logger LOGGER = LoggerFactory.getLogger(CdcStreamPartitionMetadataTask.class);
+    private static final String CDC_STREAM_STATUS_UPSERT_SQL
+            = "UPSERT INTO " + SYSTEM_CDC_STREAM_STATUS_NAME + " VALUES (?, ?, ?)";
+
+    // parent_partition_id will be null, set partition_end_time to -1
+    private static final String CDC_STREAM_PARTITION_UPSERT_SQL
+            = "UPSERT INTO " + SYSTEM_CDC_STREAM_NAME + " VALUES (?,?,?,null,?,-1,?,?)";
+
+    @Override
+    public TaskRegionObserver.TaskResult run(Task.TaskRecord taskRecord) {
+        Configuration conf = HBaseConfiguration.create(env.getConfiguration());
+        Configuration configuration = HBaseConfiguration.addHbaseResources(conf);
+        int getTableRegionsTimeout = configuration.getInt(PHOENIX_STREAMS_GET_TABLE_REGIONS_TIMEOUT,
+                DEFAULT_PHOENIX_STREAMS_GET_TABLE_REGIONS_TIMEOUT);
+        PhoenixConnection pconn = null;
+        String tableName = taskRecord.getTableName();
+        String streamName = taskRecord.getSchemaName();
+        Timestamp timestamp = taskRecord.getTimeStamp();
+        try {
+            pconn = QueryUtil.getConnectionOnServer(env.getConfiguration()).unwrap(PhoenixConnection.class);
+            List<HRegionLocation> tableRegions = pconn.getQueryServices().getAllTableRegions(
+                    tableName.getBytes(), getTableRegionsTimeout);
+            upsertPartitionMetadata(pconn, tableName, streamName, tableRegions);
+            updateStreamStatus(pconn, tableName, streamName);
+            return new TaskRegionObserver.TaskResult(TaskRegionObserver.TaskResultCode.SUCCESS, "");
+        } catch (Throwable t) {
+            LOGGER.error("Exception while bootstrapping CDC Stream Partition Metadata for "
+                    + taskRecord.getTableName() + " and timestamp " + timestamp.toString(), t);
+            return new TaskRegionObserver.TaskResult(TaskRegionObserver.TaskResultCode.FAIL, t.toString());
```

Comment:
TODO: Consider a RETRY strategy here.

Reply:
@haridsv @virajjasani Not sure under what kind of failures, we should be retrying the task here?

Reply:
done